lib/web/recordingplayback.go

```diff
@@ -0,0 +1,677 @@
+		return nil, trace.BadParameter("missing session ID in request URL")
+	}
+
+	defer ws.Close()
+
+	ctx := r.Context()
+	clt, err := sctx.GetUserClient(ctx, cluster)
+	if err != nil {
+		data := []byte(err.Error())
+
+		totalSize := responseHeaderSize + len(data)
+		buf := make([]byte, totalSize)
+
+		encodeEvent(buf, 0, eventTypeError, 0, data, 0)
+
+		if err := ws.WriteMessage(websocket.BinaryMessage, buf); err != nil {
+			h.logger.ErrorContext(ctx, "failed to send event", "session_id", sessionID, "error", err)
+		}
+
+		return nil, nil
+	}
+
+	playback := newRecordingPlayback(ctx, ws, clt, sessionID, h.logger)
+
+	playback.run()
+
+	return nil, nil
+}
+
+// newRecordingPlayback creates a new session recording playback handler.
+// This provides a way for the client to request session events within a specific time range, as well as the current
+// terminal screen state at a given time (when seeking).
+// This allows for faster seeking without having to send the client extra events to reconstruct the terminal state.
+func newRecordingPlayback(ctx context.Context, ws *websocket.Conn, clt events.SessionStreamer, sessionID string, logger *slog.Logger) *recordingPlayback {
+	ctx, cancel := context.WithCancel(ctx)
+
+	s := &recordingPlayback{
+		ctx:       ctx,
+		cancel:    cancel,
+		clt:       clt,
+		sessionID: sessionID,
+		logger:    logger,
+		ws:        ws,
+		writeChan: make(chan []byte),
+	}
+
+	return s
+}
+
+// run starts the recording playback handler.
+func (s *recordingPlayback) run() {
+	defer s.cleanup()
+
+	go s.writeLoop()
+
+	s.readLoop()
+}
+
+// cleanup cleans up the recording playback resources.
+func (s *recordingPlayback) cleanup() {
+	s.cancel()
+
+	_ = s.ws.Close()
+}
+
+// writeLoop handles all websocket writes from a dedicated goroutine.
+func (s *recordingPlayback) writeLoop() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case msg, ok := <-s.writeChan:
+			if !ok {
+				return
+			}
+
+			if err := s.ws.SetWriteDeadline(time.Now().Add(10 * time.Second)); err != nil {
+				s.handleWebsocketError(err)
+				return
+			}
+
+			if err := s.ws.WriteMessage(websocket.BinaryMessage, msg); err != nil {
+				s.handleWebsocketError(err)
+				return
+			}
+		}
+	}
+}
+
+// handleWebsocketError handles errors that occur during websocket writes.
+func (s *recordingPlayback) handleWebsocketError(err error) {
+	if !websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) &&
+		!utils.IsOKNetworkError(err) {
+		s.logger.ErrorContext(s.ctx, "websocket write error", "error", err)
+	}
+}
+
+// readLoop reads messages from the websocket connection and processes them.
+func (s *recordingPlayback) readLoop() {
+	for {
+		msgType, data, err := s.ws.ReadMessage()
+		if err != nil {
+			s.handleWebsocketError(err)
+			return
+		}
+
+		if msgType != websocket.BinaryMessage {
+			s.logger.DebugContext(s.ctx, "ignoring non-binary websocket message", "session_id", s.sessionID, "type", msgType)
```

Comment:
should we mark this as an hard error? The protocol requires binary communication and yet the client sent something else